src/codegen.cpp

```diff
@@ -681,6 +712,124 @@ static Function *to_function(jl_lambda_info_t *li)
     return f;
 }
 
+// TODO: a lot of duplication with to_function and init_julia_llvm_env
+// TODO: invoke by call()ing a @target("ptx") function
+extern "C" DLLEXPORT
+const jl_value_t * jl_to_ptx() {
```

Comment:
So after taking a look at https://github.com/Keno/DIDebug.jl it should be possible to write `jl_to_ptx` in terms of `CXX.jl` and keep it out of Julia base.

This would slim down this PR to `@target` (and maybe `@kernel`), some special cases for NVPTX, and more or less independent PRs for `preserving_bitcast` and `llvmcall`